### PR TITLE
Grid integration (double loop) modeling extension

### LINKED ISSUE #432: Grid Integration: double-loop generalization
Timeline:

07/23/2021

- [x] Finish generalization of MIP model objects
- [x] Add documentation
- [x] Ask @adowling2  for review

07/30/2021

- [x] MILP with and without energy storage running
- [x] Add documentation
- [x] Ask @jsiirola  and @Robbybp  for review 

08/06/2021

- [x] Rankine cycle model from Jaffer running

08/13/2021

- [ ] Finish documentation
- [ ] PR under review

08/20/2021

- [ ] Reply to all reviews
- [ ] Last chance for input

08/23/2021

- [ ] Merge PR

Future enhancements:

- Advance time

### PR DESCRIPTION
**Pull request by xiangao1**
## Changes proposed in this PR:
- generalize `Bidder`, `Tracker`, `DoubleLoopCoordinator` classes.
- Simple Rankine cycle example
- Documentation

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by xiangao1** [comment]
@jalving Please push new changes to this branch on my fork and then we will use this PR to merge into the IDAES repo.

@jsiirola @adowling2 We will ping you when this is ready for review.

**Comment by adowling2** [comment]
@ksbeattie We are targeting the February/March/April release with this.

**Comment by xiangao1** [comment]
@jalving double-loop branch on my fork is ready for you to push.

**Comment by ksbeattie** [comment]
Any progress on this and/or PR #432 ?

**Comment by darrylmelander** [comment]
Alex asked me to post an example of a Prescient plugin that is entirely in memory instead of saving a jupyter code block to a file.  Here's that example.

```python
import prescient.simulator
# Example of how to set up prescient configuration. This isn't enough
# to successfully run Prescient, just giving an example...
simulator = prescient.simulator.Prescient()
options = {'start_date':'07-10-2020',
           'num_days': 7,
           'sced_horizon':4,
          }
simulator.config.set_value(options)

### This is code for an example plugin. We'll create the plugin module
#   in memory, without using the file system.
#   The line below is what we want to avoid...
#%%writefile my_plugin.py

from pyomo.common.config import ConfigDict, ConfigValue
import prescient.plugins as pplugins
from prescient.simulator.config import PrescientConfig

def get_plugin_configuration(key):
    config = ConfigDict()
    config.declare('print_callback_message', 
                   ConfigValue(domain=bool,
                               description='Print a message when callback is called',
                               default=False)).declare_as_argument(f'--{key}.print-callback-message')
    return config

def register_plugin_points(context: pplugins.PluginRegistrationContext,
                           options: PrescientConfig,
                           plugin_config: ConfigDict) -> None:

    # print a message at initialization time
    def initialization_callback(options, simulator):
        if plugin_config.print_callback_message:
            print('  *** Called plugin function during initialization ***')
    context.register_initialization_callback(initialization_callback)

# End of plugin code
#######

# Turn the functions above into an in-code module
import types
plugin_module = types.ModuleType('my_plugin')
plugin_module.get_configuration = get_plugin_configuration
plugin_module.register_plugins = register_plugin_points

# Register the plugin with the Prescient simulator
simulator.config.plugin['my_plugin'] = plugin_module

# Our example has a plugin-specific configuration setting. Set it here.
simulator.config.plugin['my_plugin'].print_callback_message = True

# Run prescient.
# This example isn't set up to run correctly, but it will 
# get far enough to show the plugin's print statement.
simulator.simulate()
```

**Comment by adowling2** [comment]
@xiangao1 What needs to happen to get this ready to review? Can you create a checklist here? Then tag folks when it is ready to review.

**Comment by lbianchi-lbl** [comment]
@xiangao1 I've noticed that the documentation being added by this PR has its own `docs` folder and `conf.py` file. While this might be convenient while developing locally, it's not integrated with the IDAES documentation infrastructure, and would therefore be left out from https://idaes-pse.readthedocs.org.

Before being able to merge this, we want to make sure the docs are properly integrated. Off the top of my head, this includes:

- [ ] Move the `.rst` files (plus additional material such as e.g. images) from the current location `idaes/apps/grid_integration` to the top-level `docs` directory alongside to the docs for the rest of `idaes.apps`, i.e. `docs/explanations/modeling_extensions/grid_integration`
- [ ] Make sure that the docs build correctly and all the pages (including, again, any additional material such as e.g. images) show up in the general IDAES docs after building
- [ ] Remove the `docs` directory and all redundant configuration or build files (e.g. `conf.py`, Makefiles, ...) from the `idaes/apps/grid_integration`

To view the HTML docs, you can either:

- Build the HTML documentation locally by running `python build.py` from the `docs` folder, then opening any of the build `.html` files with a web browser
- Access the online version that ReadTheDocs builds as part of the CI checks for every PR:
  
image

Similarly, we should make sure that any `.ipynb` file will be part of the examples repository at https://github.com/IDAES/examples-pse.

**Comment by xiangao1** [comment]
> @xiangao1, any progress on this?

@ksbeattie Thanks for checking in. I am working on it.

**Comment by xiangao1** [comment]
Hi @adowling2, could you review this? Thanks!

**Comment by ksbeattie** [comment]
@xiangao1, could you take a look at resolving the conflicts here why waiting for @adowling2's review?

**Comment by adowling2** [comment]
@ksbeattie I do not think @xiangao1's edits should have touched these two conflicted files. Is there an easy way to "stash" these changes and then just revert to what is in the main branch for these two files?

**Comment by adowling2** [comment]
@xiangao1 Do we have an issue with a checklist for changes that need to get done before this is marked as ready to review?

File: idaes/apps/grid_integration/multiperiod/dev.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#################################################
```
Comment (adowling2) [inline idaes/apps/grid_integration/multiperiod/dev.py:1]: Is this file needed? `dev.py` sounds like something that was used for testing, etc., but should not be part of the release. If it is important, let's add a more meaningful comment block at the top.

**Review by adowling2 — COMMENTED** [review]
This looks good. @Robbybp Any thoughts on why this PR is modifying caprese files?

**Comment by ksbeattie** [comment]
> @ksbeattie I do not think @xiangao1's edits should have touched these two conflicted files. Is there an easy way to "stash" these changes and then just revert to what is in the main branch for these two files?

Hmm, this is concerning.  @xiangao1, can you confirm if you expected changes to these (and lots of other caprese) files in this PR?

I just noticed that the branch for this PR is coming from Xian's fork which is forked from @Robbybp's fork, not directly from the upstream idaes-pse repo.  I'm worried that because of this we might be seeing "extra changes" in this PR, like these conflicts.

@xiangao1 you should make sure that the list of files changed, and all the changes in each, are **all** of the changes - **and only** the changes - you expect for this PR.  Otherwise, we might need to re-create the PR from a new, up-to-date fork or branch.

**Comment by andrewlee94** [comment]
@xiangao1 Just quickly looking over the files myself, I noticed a folder named `idaes/apps/rankine` with only two zipped files as the contents. Is this meant to be part of this PR, and if so, what is this exactly?

The second thing I noticed is that there are no docs yet. We would want to see at least some basic docs before we merge this, at a minimum describing what the tools is for and a basic example of how to use it.

**Comment by xiangao1** [comment]
@ksbeattie The reason why I have changes on Caprese files is because I merged with Robby's fork, and those changes on Caprese have not been merged into IDAES main. Should I create a new PR and add my changes there? Thanks!

**Comment by xiangao1** [comment]
> @xiangao1 Just quickly looking over the files myself, I noticed a folder named `idaes/apps/rankine` with only two zipped files as the contents. Is this meant to be part of this PR, and if so, what is this exactly?
> 
> The second thing I noticed is that there are no docs yet. We would want to see at least some basic docs before we merge this, at a minimum describing what the tools is for and a basic example of how to use it.

@andrewlee94 In the `idaes/apps/rankine`, there are files written by @jalving and @jghouse88 for the simple Rankine cycle model. And this model is used for the examples in the `idaes/apps/grid_integration`. Should we create another PR for this simple Rankine cycle model?

I do have docs written for this package here: https://github.com/xiangao1/idaes-pse/tree/double-loop/idaes/apps/grid_integration/docs. Should I move them to somewhere else? Thanks!

**Comment by adowling2** [comment]
@ksbeattie @lbianchi-lbl Is there a "proper GitHub way" for @xiangao1 to move only changes to specific files he designates to a new branch so as to not include the unmerged changes on @Robbybp's fork while maintaining the git history of the edited files (the ones we want to keep)?

**Comment by adowling2** [comment]
> @xiangao1 Just quickly looking over the files myself, I noticed a folder named `idaes/apps/rankine` with only two zipped files as the contents. Is this meant to be part of this PR, and if so, what is this exactly?
> 
> The second thing I noticed is that there are no docs yet. We would want to see at least some basic docs before we merge this, at a minimum describing what the tools is for and a basic example of how to use it.

There are some docs in `idaes/apps/grid_integration/docs/`. Will these get compiled into readthedocs when the PR is not marked as Draft?

**Comment by adowling2** [comment]
@andrewlee94 Good suggestion. @jghouse88 is the main developer of the simple Rankine cycle, so let's revisit this is a few weeks. I like your suggestion of making it a separate PR.

**Comment by adowling2** [comment]
Discussion topic for the dev call: Close this PR in favor of #787
